Stream Claim Id

underhood opened the pull request:
##### Summary
claimed_id of nodes is streamed to parents. For easy testing also added to `/api/v1/info` (where cloud will need it anyway sooner or later)
Example of 3-level streaming from the view of top-level parent:
![Screenshot from 2020-08-25 11-38-41](https://user-images.githubusercontent.com/6674623/91158903-9a4f7380-e6c7-11ea-8c7d-07491b08c330.png)


##### Component Name
Streaming
##### Test Plan
Have some streaming setup and look at `/api/v1/info`. Check nothing breaks with Cloud by connecting this node to `staging.netdata.cloud`. Claim nodes and see if their claiming Id propagates up the network by checking `/api/v1/info` of the topmost node. All nodes have to be running this PR for it to work.

What was tested together with the cloud team:
1. Claim node to Staging and connect it. Node had 2 children.
1. Message as per screenshot above (with new `mirrored_hosts_status` field) is send to cloud as part of `onconnect` payload over ACLK
1. Checked Cloud GUI still works as expected with this new fields
1. Cloud team checked there are no errors in the background (not visible to us from GUI)

What was tested additionally:
- node that is claimed during runtime and then `netdatacli reload-claiming-state`
- up to 4 level hierarchy of streaming
    - `TopLevelMaster` (VM,not claimed)
    - `timowss` (real machine,claimed) -> `TopLevelMaster` (VM)
    - `Fed32-Layer1Host1` (VM,claimed) -> `timowss` (real machine, claimed)
    - `Fed32-Layer1Host2` (VM,not claimed) -> `timowss` (real machine, claimed)
    - `Fed32-Layer2Host1` (VM,not claimed) -> `Fed21-Layer1Host1` (VM, claimed)
    - this represents final test with biggest amount of hosts tested, during development and testing different combination of nodes and claiming states were used
- both claimed and unclaimed nodes have correct status reported

##### Additional Information
Codacy false triggers here. It doesn't seem to recognize `rrdhost_foreach_read` macro is a loop and as a result, thinks `if(count > 0) {` is always false (which is not true).

amoss commented on web/api/web_api_v1.c:
```diff
@@ -775,13 +775,44 @@ static inline void web_client_api_request_v1_info_summary_alarm_statuses(RRDHOST
 
 static inline void web_client_api_request_v1_info_mirrored_hosts(BUFFER *wb) {
     RRDHOST *rc;
+    BUFFER *guids = buffer_create(2048);
```
This buffer is allocated and freed only so that there is a single loop over the hosts. But looping over the hosts is cheap - it is only walking down a linked list. Allocating and freeing the buffer is expensive as it involves syscalls. Why not two loops that both `buffer_sprintf(wb` instead of two buffers?

underhood replied:
OK I will change it. Yep it was done to go trough the list only once while lock is held. There is therefore small improvement during lock but as you correctly pointed out at cost outside of the critical section. This would be improved also by mempooling the buffers but that is another topic

amoss commented on streaming/receiver.c:
```diff
@@ -95,6 +95,54 @@ PARSER_RC streaming_timestamp(char **words, void *user, PLUGINSD_ACTION *plugins
+    rrdpush_claimed_id(host);
+
+    return PARSER_RC_OK;
+}
```
The error messages in this function are not useful. If this breaks then we need to diagnose it by checking logs, so at the point that an error is encountered there needs to be enough information in the log to pinpoint the problem.
* If the claim ID string is not a valid form then show the string.
* If the GUID is not valid then show the string.
* If the child is attempting to change the wrong UUID then show the incorrect UUIDS.
This does not jut help us - users with broken config will look in their logs to try to understand the problem and it should contain enough information.

The conversation above has been marked resolved - but the problem is not resolved. A child cannot change the claim id of an arbitrary host. This means that configuration errors propagate through the system and break the state on other nodes. Find another way to handle it,

underhood replied:
I will improve the logging. 
Host must be able to modify claim_id of another host than self (as it can itself have children). Not sure yet how to improve this other than adding visibility of structure of children below (as a tree instead of list)

amoss replied:
The terminology is a bit confusing here (historically I think the terms have shifted a bit). When a sender connects to a receiver it only transmits a single *host* - in this context the host is a particular database. When a proxy node is transmitting the data for multiple children to a parent there one stream per host. So the protocol that is coming over the wire is only describing a single host (even though the node may be streaming multiple hosts over multiple connections).

underhood replied:
ah I see what you mean now. Knew there is connection per host but somehow misunderstood/didn't make a connection with what you try to say.

Changed it. Thanks

amoss commented on streaming/sender.c:
```diff
@@ -622,6 +622,15 @@ void *rrdpush_sender_thread(void *ptr) {
+                        rrdpush_claimed_id(h);
+                }
+                rrd_unlock();
+            }
```
When an agent is streaming data from multiple hosts each one has a separate connection and sender.

underhood replied:
you are right done

amoss reviewed (changes requested):
1. Improve the logging.
2. Fix the loop with the unnecessary buffer
3. Find a better solution for updating claim ids on other hosts that does not permit the system to be broken.
4. Either explicitly confirm with the cloud team that the changes do not break anything or put in the version negotiation first and change the protocol version.

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6051564&bid=19824351)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6051564&bid=19824351)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6051564&bid=19824351)

underhood commented:
@amoss in regards to point 4 it is confirmed with cloud team now this can go in. the `hello` version negotiation message (this will be another PR) will need separate explicit GO from cloud team

amoss commented:
The test plan is insufficient. This needs to be tested it by connecting it to staging and checking with the cloud team that the data received is in a format that they can process without errors.

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6051564&bid=19824351)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6051564&bid=19824351)

underhood commented:
@amoss all your points are resolved now.
1. done
2. I changed it to your wishes: so it makes critical section a bit longer (loop over hosts twice instead of once) but it removes the expense of Netdata BUFFER allocation
3. It would possibly break system only if child had a bug but it is good suggestion and I added additional check on what child sends.
4. clarified with cloud team and have explicit approval for `/api/v1/info` payload update that is done in this PR

amoss commented:
The test plan does not describe the testing in enough detail for a reviewer to decide if this is properly tested. Please fill it in properly:
1. What was the test scenario
2. What was sent to the cloud
3. What was checked on the far end

underhood commented:
made test plan more verbose

amoss commented:
What was the 4-level setup? Which nodes were the parents of which nodes? Which nodes were claimed and which nodes were not claimed?

underhood commented:
updated info on 4 layer stream test performed

mfundul commented on web/api/web_api_v1.c:
```diff
@@ -776,14 +776,57 @@ static inline void web_client_api_request_v1_info_summary_alarm_statuses(RRDHOST
+            continue;
+        if (count > 0)
+            buffer_strcat(wb, ",\n");
+
```
I don't think we should nest those locks, they are completely unrelated and I don't like the implications. We could just copy the strings somewhere or maybe use `buffer_strcat()`?

underhood replied:
avoided holding both locks at expense of strdupz

underhood commented:
@amoss 
![ZO_12i8m38RlVOgm-mwjDr7OIJpCcoyGjuYJr4gJAO8FhyLDN57myfU4VlnVAb6CsdisONUaHc4RpZLTY3SeIj5CHvX3mzXvQc49UsIX4IjMrCvSMUEDebq7KLi02h0zM9DYGchitw1sFbn5pF3c_tuNkB70cHJ_gRaFDVTKI_65pCzUtCioDtqjqstPPZ2omvXgAZgrlUS7](https://user-images.githubusercontent.com/6674623/91162521-78a4bb00-e6cc-11ea-8691-d9de17a64bb9.png)

this is topology of the 4 layer test However during testing I tried different combinations and topologies

underhood commented on claim/claim.c:
```diff
@@ -27,18 +27,16 @@ static char *claiming_errors[] = {
-    result = (claimed_id == NULL) ? NULL : strdup(claimed_id);
-    netdata_mutex_unlock(&claim_mutex);
+    netdata_mutex_lock(&localhost->claimed_id_lock);
+    result = (localhost->claimed_id == NULL) ? NULL : strdup(localhost->claimed_id);
```
thx, comes form pre-existing code but you are right and I will fix it

underhood commented on streaming/receiver.c:
```diff
@@ -95,6 +95,53 @@ PARSER_RC streaming_timestamp(char **words, void *user, PLUGINSD_ACTION *plugins
+        return PARSER_RC_ERROR;
+    }
+
+    // We don't need to parsed UUID
```
thanks

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6051564&bid=19824351)

amoss commented:
Ok good, now the test plan is comprehensive. I'll let the other reviewer take a look before I go back for a second look at the code.

amoss commented:
I've disabled the Codacy warnings as they are just noise - it does not understand the loop macro for some reason.

mfundul commented on web/api/web_api_v1.c:
```diff
@@ -776,14 +776,57 @@ static inline void web_client_api_request_v1_info_summary_alarm_statuses(RRDHOST
+    buffer_strcat(wb, "\n\t],\n");
+}
+
+static inline void web_client_api_request_v1_info_mirrored_hosts_uids(BUFFER *wb) {
```
Is this called somewhere? Cause I can't find it.

underhood replied:
aaah artifact from cherry-picked commit from another branch
removed, thanks

amoss commented on streaming/rrdpush.c:
```diff
@@ -364,6 +364,21 @@ void rrdpush_send_labels(RRDHOST *host) {
+    buffer_sprintf(host->sender->build, "CLAIMED_ID %s %s\n", host->machine_guid, (host->claimed_id ? host->claimed_id : "NULL") );
+
+    netdata_mutex_unlock(&host->claimed_id_lock);
+    sender_commit(host->sender);
```
After writing into the sender buffer the sender thread should be signalled over the pipe so that it knows there is data waiting.

amoss replied:
There is an example on https://github.com/netdata/netdata/blob/9ecb4756d7104a2dd11e74ecec31f413035c69bb/streaming/rrdpush.c#L362

underhood replied:
added

amoss commented on streaming/receiver.c:
```diff
@@ -95,6 +95,53 @@ PARSER_RC streaming_timestamp(char **words, void *user, PLUGINSD_ACTION *plugins
+        return PARSER_RC_OK; //the message is OK problem must be somewehere else
+    }
+
+    if (host == localhost) {
```
This check is redundant now, as it is captured by the check above.

amoss commented on web/api/web_api_v1.c:
```diff
@@ -776,16 +776,46 @@ static inline void web_client_api_request_v1_info_summary_alarm_statuses(RRDHOST
+    count = 0;
+    rrdhost_foreach_read(rc)
+    {
+        if (rrdhost_flag_check(rc, RRDHOST_FLAG_ARCHIVED))
```
Can you check with product if this is correct? Are they expecting child node databases to be available after a restart on the parent node?

underhood replied:
This is unchanged in current PR archived hosts were skipped before as well.
In case we need archived hosts in info payload we will need to update also the metadatalog to keep and comapct the claim_id. That however would not be 1hr task

mfundul replied:
It was specifically asked to not expose archived host by Chris.

amoss commented on web/api/web_api_v1.c:
```diff
@@ -776,16 +776,46 @@ static inline void web_client_api_request_v1_info_summary_alarm_statuses(RRDHOST
+
+    buffer_strcat(wb, "\n\t],\n\t\"mirrored_hosts_status\": [\n");
+    count = 0;
+    rrdhost_foreach_read(rc)
```
This should not be called `rc` it should be called `host`. The part that you copied it from is wrong, the contributor was working on alarms so they reused the iterator name for alarm entries.

underhood replied:
will change the variable name for clarity

underhood commented:
@Ferroin arch CI build started failing yesterday with ` configure: error: libuv required but not found. Try installing 'libuv1-dev' or 'libuv-devel'.` Can you take a look if you have 5'

underhood commented:
I checked also with production that it doesnt kill something just to be paranoid before merge.
